storage/src/nodestore/persist.rs

```diff
@@ -247,3 +364,207 @@ impl NodeStore<Arc<ImmutableProposal>, FileBacked> {
+        let mut iter = UnPersistedNodeIterator::new(&store);
+
+        assert!(iter.next().is_none());
+    }
+
+    #[test]
+    fn test_single_leaf_node() {
+        let leaf = create_leaf(&[1, 2, 3], &[4, 5, 6]);
+        let store = create_test_store_with_root(leaf.clone());
+        let mut iter =
+            UnPersistedNodeIterator::new(&store).map(|node| node.as_shared_node(&store).unwrap());
+
+        // Should return the leaf node
+        let node = iter.next().unwrap();
+        assert_eq!(*node, leaf);
+
+        // Should be exhausted
+        assert!(iter.next().is_none());
+    }
+
+    #[test]
+    fn test_branch_with_single_child() {
+        let leaf = create_leaf(&[7, 8], &[9, 10]);
+        let branch = create_branch(&[1, 2], Some(&[3, 4]), vec![(5, leaf.clone())]);
+        let store = create_test_store_with_root(branch.clone());
+        let mut iter =
+            UnPersistedNodeIterator::new(&store).map(|node| node.as_shared_node(&store).unwrap());
+
+        // Should return child first (depth-first)
+        let node = iter.next().unwrap();
+        assert_eq!(*node, leaf);
+
+        // Then the branch
+        let node = iter.next().unwrap();
+        assert_eq!(&*node, &branch);
+
+        assert!(iter.next().is_none());
+
+        // verify iterator is fused
+        assert!(iter.next().is_none());
+    }
+
+    #[test]
+    fn test_branch_with_multiple_children() {
+        let leaves = [
+            create_leaf(&[1], &[10]),
+            create_leaf(&[2], &[20]),
+            create_leaf(&[3], &[30]),
+        ];
+        let branch = create_branch(
+            &[0],
+            None,
+            vec![
+                (1, leaves[0].clone()),
+                (5, leaves[1].clone()),
+                (10, leaves[2].clone()),
+            ],
+        );
+        let store = create_test_store_with_root(branch.clone());
+
+        // Collect all nodes
+        let nodes: Vec<_> = UnPersistedNodeIterator::new(&store)
+            .map(|node| node.as_shared_node(&store).unwrap())
+            .collect();
+
+        // Should have 4 nodes total (3 leaves + 1 branch)
+        assert_eq!(nodes.len(), 4);
+
+        // The branch should be last (depth-first)
+        assert_eq!(&*nodes[3], &branch);
+
+        // Children should come first - verify all expected leaf nodes are present
+        let children_nodes = &nodes[0..3];
+        assert!(children_nodes.iter().any(|n| **n == leaves[0]));
+        assert!(children_nodes.iter().any(|n| **n == leaves[1]));
+        assert!(children_nodes.iter().any(|n| **n == leaves[2]));
+    }
+
+    #[test]
+    fn test_nested_branches() {
+        let leaves = [
+            create_leaf(&[1], &[100]),
+            create_leaf(&[2], &[200]),
+            create_leaf(&[3], &[255]),
+        ];
+
+        // Create a nested structure: root -> branch1 -> leaf[0]
+        //                                -> leaf[1]
+        //                                -> branch2 -> leaf[2]
```

Comment:
The description does not seem to match with test. 
```suggestion
        // Create a nested structure: root -> leaf[0]
        //                                -> leaf[1]
        //                                -> inner_ branch -> leaf[2]
        //                                -> persisted_node
```